0026-an-extensible-db-design.adoc

```diff
@@ -0,0 +1,181 @@
+
+==== 2 NF:
+* all attributes (non-key columns) depend on the key - if not, then a conjunction table should be created consisting out of 2 keys joining two otherwise not dependent tables (e.g. protocol and swap)
+
+==== 3 NF:
+* all fields (all columns) can be determined only by the key in the table and no other column. E.g. if for every Han_BTC swap we have the same expiries, then this should be an own table. --> update anomaly
+
+==== BCNF
+* any part of a primary key must not have a functional dependency on another column of the table.
+
+==== 4 NF:
+* no multi-valued dependencies. E.g. a customer who bought 2 items should not be in a table twice with the bought item in one column
+
+
+== Approach
+
+. Data collection: Collect all the data to be stored.
+. First model: Create a first model to act upon.
+. Iterations: Iterate on the model to improve and normalize it.
+. Final model
+
+Constraints we considered besides the normalization constraints:
+
+. FK  relationships should not be defined in a parent table (problem of not being able to create the parent if the child does not exist)
+. Certain data of a swap may only be known to a specific role at a later point in time (after creation of the swap). We want to avoid table updates.
+
+
+== Data collection
+
+We identified the following data sources needed:
+
+. Initial data of each swap when being created for a specific role for a specific alpha and beta to protocol where alpha and beta is one of: `HAN`, `HErc20`, `HALight`
+. Additional data added to the swap for a specific role (e.g. secret hash only known to Bob during announcement)
+. Additional data added to the swap's alpha or beta ledger data.
+
+While not necessarily in scope we considered the following as well:
+
+. Swap State (one per swap, aggregated state - record all states with timestamp)
+. Communication State (one per swap - record all states with timestamp)
+. Ledger State (two per swap, record all states with timestamp and additional information if applicable, e.g. transaction ID)
+
+
+== Final model
+
+Instead of starting from scratch, we start iterating over a slightly adapted table layout based on the one proposed by @bonomat
+https://3.basecamp.com/4403044/buckets/16118249/messages/2583518881#__recording_2600021433[in basecamp].
+
+The final model can be found at the end of this document.
+Our reasoning for creating certain table relations is summed up in the sections below.
+
+=== Swaps
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/14e942f9c5ea1791f01bb65db0e09ac11cf5ab4c/assets/db_schema/swap.puml&fmt=svg[Swaps]
+
+Our first challange was depicting `alpha` and `beta` protocols for the swap.
+Our first attempt was to not indicate this in the `swap` table but instead just use foreign key relationships.
+For that to work we would need to indicate in the child table whether it is `alpha` or `beta` (see below in `swap_protocol`).
```

Comment:
> That's as type safe as it gets right? :)

Not really because there is no relation between:

```
.inner_join(
han_bitcoin::table.on(han_bitcoin::local_swap_id.eq(swaps::local_swap_id))
```
and
```
.filter(swaps::alpha_protocol.eq("han_bitcoin"))
```

There is not enforcement that the data for a particular protocol is stored in the table that the `alpha_protocol` column in the swaps table suggests. Or am I missing something?

> I think the proposed approach is more elegant as otherwise there is an explosion of tables, i.e for each new supported protocol you will need to add a bunch of new tables vs right now we have 1.

Depends on how you define elegance I guess.
To me, an elegant design makes it impossible to express things that you want to be impossible when you run the program. It may result in more tables but they all really just have a couple of columns with foreign keys and no data is duplicated.

Reply:
> There is not enforcement that the data for a particular protocol is stored in the table that the alpha_protocol column in the swaps table suggests. Or am I missing something?

Do you mean that there is no fk relation between the alpha_protocol column and the table to find the swap_details? That is true indeed. We could add a fk in `{protocol}_swap_detail` to `protocol`. I'm not sure if this adds any more guarantees though as we need to ensure this when inserting the data and the table name depicts already what data should be stored in it. 

> Depends on how you define elegance I guess.
To me, an elegant design makes it impossible to express things that you want to be impossible when you run the program. It may result in more tables but they all really just have a couple of columns with foreign keys and no data is duplicated.

As far as I can see the type guarantees are the same for both approaches as in code we need to guarantee that the correct data is stored in the correct tables no?
I might be missing something here so please help me out :)

Reply:
> As far as I can see the type guarantees are the same for both approaches as in code we need to guarantee that the correct data is stored in the correct tables no?

Without fk, you are free to store whatever "protocol" you want in the swaps column and store the protocol data somewhere else, i.e. I can say "this swap is han-etheruem<->halight" in the swap table but actually store entries in the "han_bitcoin" and "herc20" protocol tables for this swap id.

Foreign-key relations in SQL (I hope in SQLite too) are _typed_, meaning a foreign key in a specific column can only point to one particular table and not any other.

Meaning if we create a schema like:

- `swap`
	- id
	- ...

- `han_ethereum_protocol`
	- id
	- wei_quantity
	- expiry
	- ...

- `han_bitcoin_protocol`
	- id
	- satoshi_quantity
	- expiry
	- ...

- `han_ethereum_han_bitcoin_swap`
	- id
	- alpha_ledger_fk -> `han_ethereum_protocol`
	- beta_ledger_fk -> `han_bitcoin_protocol`
	- swap_fk -> `swap`
	- ...

The database schema guarantees that entries in the `han_ethereum_han_bitcoin_swap` table always point to data of the actual protocols in use. In a way you avoid repeating yourself in saying what protocols are in use by just pointing to the relevant entries in the individual protocol tables. Due to the nature of typed FKs, this requires a dedicated table for each combination.

The database is going to prevent you from inserting invalid data because values in the `beta_ledger_fk` column MUST point to an entry in `han_bitcoin_protocol`, otherwise the SQL statement is not valid :)
You cannot make it point to a different protocol table (I hope that SQLite also enforces this, at least bigger database engines like PostgreSQL do that)

The downside of this is that we have one table per combination. They would only be join tables though to tie things together and not duplicate all the columns like we do in the current schema.

What we will need in any case (but that is independent of both solutions) is a way to determine the protocol combination of a swap with just the ID (in the code we currently call it `DetermineTypes`). This can be built on top of both solutions, with varying degrees of efficiency.

In summary, I am slightly preferring the solution with more safety (i.e. dedicated join tables) because that is what these tools are for. If we don't care about this kind of relational safety, then we are not getting as much out of SQL as we could and hence the argument of using it over a NoSQL solution becomes weaker.

Let me know what you think :)

Reply:
Awesome writeup. 

I think I get it, let me try to repeat: 

In the current proposal we have 

- `protocol`
	- id

- `swap`
	- id
	- alpha_protocol_fk: -> `protocol.id`
	- beta_protocol_fk: -> `protocol.id`
	- ...

- `han_ethereum_swap_details`
	- id
	- wei_quantity
	- expiry
	- ...

- `han_bitcoin_swap_details`
	- id
	- satoshi_quantity
	- expiry
	- ...

And since `han_ethereum_swap_details` and `han_bitcoin_swap_details` are not connected to the `protocol` we lose some type guarantees when wanting to load a particular swap.

I still think the type safetyness is the same despite the lack of this relationship due to how data is inserted. The database is going to ensure that the types in `han_ethereum_swap_details` and `han_bitcoin_swap_details` are of the correct type or better, the ORM will take care of this, e.g. we do `Swap.save()`, `HanBitcoinSwapDetails.save()` etc. And this is the important thing to ensure. 
So we need to ensure that the correct data will be stored in the code during insert time anyways. 
If we can't rely on this then in your proposal we would have a similar problem in a way that one can store multiple `{protocol}_swap_detail` combinations and wouldn't know which one to load (e.g. we store btc<>eth and ln_eth to with the same swap_id.

The reason why we introduced this `protocol` table and the `alpha/beta_protocol` column into `swap` was that we need to know what ledgers/protocols are involved in a swap by looking at the swap. 

I wanted to propose to just store the `protocol.id` in `{protocol}_swap_details` as well which should then fulfill your initial proposal but I realized that there is a problem with our proposal:
same chain swaps are not doable as we would not be able to identify which entry in `han_bitcoin_swap_detail` is alpha and which beta. 
Now, we can argue that this is out of scope anyways but we considered this during the design nevertheless which let to the current proposal. 

The last problem would be possible with your proposal by introducing a table `han_bitcoin_han_bitcoin_swap`.

Reply:
I propose we hop on a call to discuss this further?

Reply:
> I propose we hop on a call to discuss this further?

Definitely, can you send out a calendar invite?

> If we can't rely on this then in your proposal we would have a similar problem in a way that one can store multiple {protocol}_swap_detail combinations and wouldn't know which one to load (e.g. we store btc<>eth and ln_eth to with the same swap_id.

Do you mean that two different join tables `han_ethereum_han_bitcoin_swap` and `han_bitcoin_herc20_swap` point to the same row in `swap`?

Agreed, this is a problem but I would argue that it is different one because it is not about messing up which protocols are involved with a particular swap but rather accidentally merging two swaps together. We could (partly) address that by not having a `swap` table and instead putting all the "common" fields into those tables (secret_hash, shared_swap_id, etc).

Reply:
My summary:

To add full safety we might really have to add a table for every combination of swap we allow, i.e. `han_btc_han_eth_swap`, `han_eth_han_btc_swap` (...), but that is not feasible. 

I feel adding the foreign relationship from `protocol` go `*_swap_detail` is a half baked solution, because nobody can stop you from putting a wrong protocol into a detail. 

I played around a bit with the DB tool we initially used for modelling, but could not come up with a better solution so far.

Reply:
I think modelling `alpha` and `beta` as joint tables for each ledger is a valid solution for this problem. 
This would also grow linearly, we would have 2 tables more per ledger we add.
@thomaseizinger what do you think?

![image](https://user-images.githubusercontent.com/5557790/80670057-4baf1580-8ae9-11ea-93b8-ade0c8ef6595.png)

Reply:
With @da-kami's proposed solution it is very cumbersome to know what ledgers we are swapping in between.
If we are fine with this added complexity I would opt for a solution with less tables: Instead of having 2 more tables per `{protocol}_swap_detail` table I would opt for a de-normalized solution and just add 1 field to `{protocol}_swap_detail` indicating whether an entry is alpha or beta (and remove the protocol table and `alpha`/`beta` from `Swap`). 

this allows us to do two joins on Swap to load HanBTC and HanETH swap details: 
```
swap.left_join(han_btc where side=alpha).left_join(han_eth where side = beta)
```

Reply:
Interesting idea, haven't considered that before!

That just stimulated another thought:

Nullable fk columns in the `swap` table:

- alpha_ledger_han_bitcoin -> `han_bitcoin_swap_details`
- beta_ledger_han_bitcoin -> `han_bitcoin_swap_details`
- alpha_ledger_han_ethereum -> `han_ethereum_swap_details`

etc

Have more than two of these columns set would be invalid and that could (in theory) be enforced through database triggers but let's not go there :sweat_smile: 

It would drastically reduce the number of tables AND we could determine with a single query, what the protocol combination of a swap is.

Reply:
> With @da-kami's proposed solution it is very cumbersome to know what ledgers we are swapping in between.
> If we are fine with this added complexity I would opt for a solution with less tables: Instead of having 2 more tables per `{protocol}_swap_detail` table I would opt for a de-normalized solution and just add 1 field to `{protocol}_swap_detail` indicating whether an entry is alpha or beta (and remove the protocol table and `alpha`/`beta` from `Swap`).
> 
> this allows us to do two joins on Swap to load HanBTC and HanETH swap details:
> 
> ```
> swap.left_join(han_btc where side=alpha).left_join(han_eth where side = beta)
> ```

When we were working on the spike, the assumption was, that we would know up-front which combination we want to load. So that should not be a problem.

Your solution is fine for me too, and is definitely less complex in tables.

Reply:
I thought about all three ideas and came to the conclusion that all of them are valid solutions.

Notes:
My solution builds purely on relationships.

Thomas solution maps my solution from table-relationships to table-columns. We would have more and more columns that produce more and more null-values. 

Philipps solution is base on fields in the child rather than relationships - it solves the problem through value entries of `alpha` and `beta` - the relationship would be. built on values.

Feel we covered all three levels of abstraction (table, column, field) 😄
For me it's hard to decide which one to go for. Mine is definitely the one with the highest complexity in joins 😬

Reply:
> @bonomat when we were working on the spike, the assumption was, that we would know up-front which combination we want to load. So that should not be a problem.

We have to load it from the database somehow. There is no other choice.

> Thomas solution maps my solution from table-relationships to table-columns. We would have more and more columns that produce more and more null-values.
> Philipps solution is base on fields in the child rather than relationships - it solves the problem through value entries of alpha and beta - the relationship would be built on values.

Let's go with Philipp's solution here. Seems to scale better. Nullable FK columns are kind of scary at the end of the day :D

I think with that, we've resolved all the comments! I just wanted to say good work everyone, I feel we have a solid understanding now and weighed off all the important aspects :)
I am really happy with the result!
Any chance we can get a short summary of this into the spike?

Reply:
Sounds good. I'll try to summarize this discussion in a short paragraph :)